File: tests/wasm/rust-comparison.js
/**
 * Utilities for comparing WASM binding outputs with native Rust outputs
 */
import { exec } from 'child_process';
import { promisify } from 'util';
import path from 'path';
import { fileURLToPath } from 'url';
import { existsSync } from 'fs';

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.join(__dirname, '../..');

/**
 * Get Rust output for an indicator
 * @param {string} indicatorName - Name of the indicator
 * @param {string} source - Data source (default: 'close')
 * @returns {Promise<Object>} Rust output data
 */
export async function getRustOutput(indicatorName, source = 'close') {
    
    const binaryPath = path.join(projectRoot, 'target', 'release', 'generate_references');
    const binaryExists = existsSync(binaryPath) || existsSync(binaryPath + '.exe');
    
    if (!binaryExists) {
        
        
        let buildError;
        for (let attempt = 0; attempt < 3; attempt++) {
            try {
                await execAsync('cargo build --release --bin generate_references', {
                    cwd: projectRoot,
                    env: { ...process.env, CARGO_INCREMENTAL: '0' } 
                });
                buildError = null;
                break;
            } catch (error) {
                buildError = error;
                if (attempt < 2) {
                    
                    await new Promise(resolve => setTimeout(resolve, 1000));
                }
            }
        }
        
        if (buildError) {
            throw new Error(`Failed to build reference generator: ${buildError.message}`);
        }
    }
    
    
    try {
        const { stdout, stderr } = await execAsync(
            `cargo run --release --bin generate_references -- ${indicatorName} ${source}`,
            {
                cwd: projectRoot,
                maxBuffer: 10 * 1024 * 1024 
            }
        );
        
        if (stderr) {
            console.warn(`Reference generator stderr: ${stderr}`);
        }
        
        return JSON.parse(stdout);
    } catch (error) {
        throw new Error(`Failed to generate reference for ${indicatorName}: ${error.message}`);
    }
}

/**
 * Compare WASM output with Rust output
 * @param {string} indicatorName - Name of the indicator
 * @param {Float64Array|Array} wasmOutput - Output from WASM binding
 * @param {string} source - Data source (default: 'close')
 * @param {Object} params - Parameters used
 * @param {number} tolerance - Comparison tolerance (default: 1e-8)
 * @returns {Promise<boolean>} True if outputs match
 */
export async function compareWithRust(indicatorName, wasmOutput, source = 'close', params = null, tolerance = 1e-8) {
    const rustData = await getRustOutput(indicatorName, source);
    
    
    if (indicatorName === 'damiani_volatmeter' && typeof wasmOutput === 'object' && 'vol' in wasmOutput && 'anti' in wasmOutput) {
        
        const rustVol = rustData.vol;
        const rustAnti = rustData.anti;
        
        
        if (params) {
            const rustParams = rustData.params;
            for (const [key, value] of Object.entries(params)) {
                if (rustParams[key] !== undefined && rustParams[key] !== value) {
                    throw new Error(`Parameter mismatch for ${key}: Rust=${rustParams[key]}, WASM=${value}`);
                }
            }
        }
        
        
        compareArrays(wasmOutput.vol, rustVol, indicatorName + ' vol', tolerance);
        
        compareArrays(wasmOutput.anti, rustAnti, indicatorName + ' anti', tolerance);
        
        return true;
    }
    
    if (indicatorName === 'di' && typeof wasmOutput === 'object' && 'plus' in wasmOutput && 'minus' in wasmOutput) {
        
        const rustPlus = rustData.plus;
        const rustMinus = rustData.minus;
        
        
        if (params) {
            const rustParams = rustData.params;
            for (const [key, value] of Object.entries(params)) {
                if (rustParams[key] !== undefined && rustParams[key] !== value) {
                    throw new Error(`Parameter mismatch for ${key}: Rust=${rustParams[key]}, WASM=${value}`);
                }
            }
        }
        
        
        compareArrays(wasmOutput.plus, rustPlus, indicatorName + ' plus', tolerance);
        
        compareArrays(wasmOutput.minus, rustMinus, indicatorName + ' minus', tolerance);
        
        return true;
    }
    
    
    const rustOutput = rustData.values;
    
    
    if (params) {
        const rustParams = rustData.params;
        for (const [key, value] of Object.entries(params)) {
            if (rustParams[key] !== undefined && rustParams[key] !== value) {
                throw new Error(`Parameter mismatch for ${key}: Rust=${rustParams[key]}, WASM=${value}`);
            }
        }
    }
    
    
    if (wasmOutput.length !== rustOutput.length) {
        throw new Error(`Length mismatch: WASM=${wasmOutput.length}, Rust=${rustOutput.length}`);
    }
    
    
    for (let i = 0; i < wasmOutput.length; i++) {
        const wasmVal = wasmOutput[i];
        const rustVal = rustOutput[i];
        
        
        if (isNaN(wasmVal) && isNaN(rustVal)) {
            continue;
        }
        
        const diff = Math.abs(wasmVal - rustVal);
        const tol = tolerance * (1 + Math.abs(rustVal));
        
        if (diff > tol) {
            throw new Error(
                `${indicatorName} mismatch at index ${i}: ` +
                `WASM=${wasmVal}, Rust=${rustVal}, ` +
                `diff=${diff}, tol=${tol}`
            );
        }
    }
    
    return true;
}

/**
 * Helper function to compare two arrays
 */
function compareArrays(wasmArray, rustArray, name, tolerance) {
    if (wasmArray.length !== rustArray.length) {
        throw new Error(`Length mismatch for ${name}: WASM=${wasmArray.length}, Rust=${rustArray.length}`);
    }
    
    for (let i = 0; i < wasmArray.length; i++) {
        const wasmVal = wasmArray[i];
        const rustVal = rustArray[i];
        
        
        if (isNaN(wasmVal) && isNaN(rustVal)) {
            continue;
        }
        
        const diff = Math.abs(wasmVal - rustVal);
        const tol = tolerance * (1 + Math.abs(rustVal));
        
        if (diff > tol) {
            throw new Error(
                `${name} mismatch at index ${i}: ` +
                `WASM=${wasmVal}, Rust=${rustVal}, ` +
                `diff=${diff}, tol=${tol}`
            );
        }
    }
}

/**
 * Helper to compare arrays with tolerance
 * @param {Array|Float64Array} actual - Actual values
 * @param {Array} expected - Expected values
 * @param {number} tolerance - Comparison tolerance
 * @param {string} msg - Error message prefix
 */
export function assertArrayClose(actual, expected, tolerance = 1e-8, msg = '') {
    if (actual.length !== expected.length) {
        throw new Error(`${msg}: Length mismatch: ${actual.length} vs ${expected.length}`);
    }
    
    for (let i = 0; i < actual.length; i++) {
        const diff = Math.abs(actual[i] - expected[i]);
        if (diff > tolerance) {
            throw new Error(
                `${msg}: Mismatch at index ${i}: ` +
                `expected ${expected[i]}, got ${actual[i]} (diff: ${diff})`
            );
        }
    }
}

File: tests/python/rust_comparison.py
"""Utilities for comparing Python binding outputs with native Rust outputs"""
import subprocess
import json
import numpy as np
import os
from pathlib import Path

def get_rust_output(indicator_name, source='close'):
    """Run the Rust reference generator to get expected outputs.

    Prefers a prebuilt binary to avoid Cargo lock contention during pytest. If
    the binary is not present, falls back to `cargo build` + `cargo run`.
    Honor env var `RUST_REF_BIN` when set.
    """
    project_root = Path(__file__).parent.parent.parent


    bin_hint = os.environ.get('RUST_REF_BIN', '').strip()
    candidates = []
    if bin_hint:
        candidates.append(Path(bin_hint))

    candidates.append(project_root / 'target' / 'release' / ('generate_references.exe' if os.name == 'nt' else 'generate_references'))
    candidates.append(project_root / 'target-py' / 'release' / ('generate_references.exe' if os.name == 'nt' else 'generate_references'))

    for cand in candidates:
        if cand.is_file():
            result = subprocess.run([str(cand), indicator_name, source], cwd=project_root, capture_output=True, text=True)
            if result.returncode != 0:
                raise RuntimeError(f"Failed to generate reference for {indicator_name}: {result.stderr}")
            return json.loads(result.stdout)


    build_result = subprocess.run(
        ['cargo', 'build', '--release', '--bin', 'generate_references'],
        cwd=project_root,
        capture_output=True,
        text=True
    )
    if build_result.returncode != 0:
        raise RuntimeError(f"Failed to build reference generator: {build_result.stderr}")

    result = subprocess.run(
        ['cargo', 'run', '--release', '--bin', 'generate_references', '--', indicator_name, source],
        cwd=project_root,
        capture_output=True,
        text=True
    )
    if result.returncode != 0:
        raise RuntimeError(f"Failed to generate reference for {indicator_name}: {result.stderr}")
    return json.loads(result.stdout)

def compare_with_rust(indicator_name, python_output, source='close', params=None, rtol=1e-10, atol=1e-12):
    """Compare Python binding output with native Rust output"""
    rust_data = get_rust_output(indicator_name, source)


    if isinstance(python_output, dict):

        for key, py_values in python_output.items():
            if key not in rust_data:
                raise KeyError(f"Key '{key}' not found in Rust output for {indicator_name}")
            rust_values = [float('nan') if v is None else v for v in rust_data[key]]
            rust_array = np.array(rust_values, dtype=np.float64)


            if len(py_values) != len(rust_array):
                raise ValueError(f"Length mismatch for {key}: Python={len(py_values)}, Rust={len(rust_array)}")


            try:
                np.testing.assert_allclose(py_values, rust_array, rtol=rtol, atol=atol)
            except AssertionError as e:

                for i in range(len(py_values)):
                    if np.isnan(py_values[i]) and np.isnan(rust_array[i]):
                        continue
                    diff = abs(py_values[i] - rust_array[i])
                    tol = atol + rtol * abs(rust_array[i])
                    if diff > tol:
                        raise ValueError(f"{indicator_name} {key} mismatch at index {i}: Python={py_values[i]}, Rust={rust_array[i]}, diff={diff}, tolerance={tol}")
                raise
        return True



    rust_values = [float('nan') if v is None else v for v in rust_data['values']]
    rust_output = np.array(rust_values, dtype=np.float64)


    if params:
        rust_params = rust_data['params']
        for key, value in params.items():
            if key in rust_params and rust_params[key] != value:
                raise ValueError(f"Parameter mismatch for {key}: Rust={rust_params[key]}, Python={value}")


    if len(python_output) != len(rust_output):
        raise ValueError(f"Length mismatch: Python={len(python_output)}, Rust={len(rust_output)}")


    try:
        np.testing.assert_allclose(python_output, rust_output, rtol=rtol, atol=atol, equal_nan=True)
        return True
    except AssertionError as e:

        for i in range(len(python_output)):
            if np.isnan(python_output[i]) and np.isnan(rust_output[i]):
                continue
            diff = abs(python_output[i] - rust_output[i])
            tol = atol + rtol * abs(rust_output[i])
            if diff > tol:
                raise AssertionError(
                    f"{indicator_name} mismatch at index {i}: "
                    f"Python={python_output[i]}, Rust={rust_output[i]}, "
                    f"diff={diff}, tol={tol}\n"
                    f"Original error: {str(e)}"
                )
        raise
